WordPress/openverse-frontend
# Convert the `user` store to an `ssrRef`

## Fixes


Fixes #1025 by @obulat

## Description


User store only contained one piece of data: the uuid that identified the user request. This ID should not be shared between requests, but should keep the SSR-set value on the client side, too.

I've replaced the `user` store with a composable that creates a `ssrRef` with a uuid on the server. `ssrRef` is set once, on the server, keeps the same value on the client side, and is not shared between requests. However, it can only be set from the `setup` function (setting it in global scope leads to sharing of the id between requests).

To ensure that the id is created in a component setup function, I've added conditional creation in two outer-most, user-entry components: `default` and `blank` layouts. And to make sure that it's not created twice, the id is saved to the media store after it's created. So, the id is created on SSR by one of the layout components if it is not available in the store. The user can enter the site from the index page, the search page, or other pages, and would still get an id. A limitation is that this id would be replaced if the user hard-refreshes the page.

Note: Nuxt 3 does not have `ssrRef`s, they are replaced with `useState`: https://v3.nuxtjs.org/getting-started/bridge-composition-api/#ssrref-and-shallowssrref

## Testing Instructions

Run the app, and open the Vue devtools media store. Go to localhost:8443 or localhost:8443/search/q=cat, and navigate after that. You should see that the `usageSessionId` is the same. If you click on refresh, you should see the id changing.
You can also access the app from a local network (e.g. from a phone). If you add logging, you will see that there are 2 `usageSessionId` values, one for each request.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
